pdf/differential_test/main.py

```diff
@@ -6,208 +6,178 @@
-            ex = [e if not callable(e) else e(dname) for e in ex]
-            env = os.environ.copy()
-            env.update(t.get('env', {}))
-            subprocess.check_call(ex, env=env, stdout=subprocess.PIPE,
-                    stderr=subprocess.PIPE)
+
+    tool_names = []
+    pages_by_tool = {}
+    for t in os.listdir(artifacts_root_dir):
+        tool_names.append(t)
+        for f in os.listdir(os.path.join(artifacts_root_dir, t)):
+            # Filename format is `page-%d.ext`
+            pageno = int(f.rsplit('.')[0].rsplit('-')[-1])
+            assert (t, pageno) not in pages_by_tool, f'{t} / {pageno} / {pages_by_tool}'
+            pages_by_tool[(t, pageno)] = f
+    tool_names.sort()
+    if not tool_names:
+        print(f'No upstream tools produce page images')
+
+    if html_out:
+        print(
+            '<!DOCTYPE html><html><title>Visual Differential</title><body><table border="1">'
+            '<thead><tr><th>Page</th>'
+        )
+        for t in tool_names:
+            print(f'<th>{t}</th>')
+        print('</thead><tbody>')
+
+    pageno = 1
+    rmse_max = collections.defaultdict(lambda: collections.defaultdict(float))
+    diff_max = collections.defaultdict(lambda: collections.defaultdict(float))
+    any_differential = False
+    while True:
+        existed = set()
 
         if html_out:
-            print('<!DOCTYPE html><html><title>Schizo Test</title><body><table border="1">')
+            print(f'<tr><td>Page&nbsp;{pageno}</td>')
 
-        pages_by_tool = {}
-        for f in os.listdir(dname):
-            for t in tool_names:
-                prefix = f'{t}-'
-                if not f.startswith(prefix):
-                    continue
-                pageno = int(f[len(prefix):].split('.')[0])
-                assert (t, pageno) not in pages_by_tool, f'{t} / {pageno} / {pages_by_tool}'
-                pages_by_tool[(t, pageno)] = f
+        page_images = []
+        for ti, t in enumerate(tool_names):
+            page_images.append(None)  # append placeholder
 
-        pageno = 1
-        rmse_max = collections.defaultdict(lambda: collections.defaultdict(float))
-        diff_max = collections.defaultdict(lambda: collections.defaultdict(float))
-        any_schizo = False
-        while True:
-            existed = set()
+            pagename = pages_by_tool.get((t, pageno))
+            if pagename is None:
+                if html_out:
+                    print(f'<td>{t} did not render this page</td>')
+                continue
+            existed.add(pagename)
 
-            if html_out:
-                print(f'<tr><td>Page {pageno}</td>')
+            pagepath = os.path.join(artifacts_root_dir, t, pagename)
+            # TODO requires conversion (imagemagick installed on machine)
+            # // testing before commit.
+            if not pagepath.endswith('.png'):
+                subprocess.check_call(['convert', pagepath, pagepath + '.png'])
```

Comment:
Won't this cause an issue with multiple downstream parsers? E.g., we are adding new files to the artifacts folder; if another downstream parser runs after the visual differential, it will find both the original page image and the .png version. This should probably go to a temp path rather than `pagepath + '.png'`